src/main/scala/com/lightbend/kafka/scala/streams/typesafe/ImplicitConverters.scala

```diff
@@ -0,0 +1,94 @@
+package com.lightbend.kafka.scala.streams.typesafe
+
+import com.lightbend.kafka.scala.streams.typesafe.unsafe.ConverterToTypeSafer
+import org.apache.kafka.streams.kstream._
+
+/** Conversions to keep the underlying abstraction from leaking. These allow
+  * us to always return a TS object instead of the underlying one.
+  *
```

Comment:
What do you mean by leaking of the underlying abstraction ? May be I am missing something but can you please point what advantage do we get out of these implicit classes instead of using the plain old implicit conversions as in https://github.com/lightbend/kafka-streams-scala/blob/develop/src/main/scala/com/lightbend/kafka/scala/streams/ImplicitConversions.scala#L19 ? 

In the implementation of `TSKStream` you pass the `unsafe` Java abstraction, call the relevant method on it and then again do a `.map(_.safe)`. Why not return the result itself and let the implicit conversion take care of converting it into the safe version ?

Reply:
It may be unfortunate to put that comment in the scaladoc of the implicits, since the library user can still use those to wrap their `KStream` instances into `TSKStream`s. With that I mean that we want the TSKStream definitions to always return a TSKType value and not a KStream (or others) directly. It's essentially the same you were already doing in KStreamS too.

What do you mean by *plain old implicit conversions*, using `implicit def` instead of an `implicit class` with a method?

I did that because it feels less error-prone to miss out a type when you change the types explicitly. In your example, if I couldn't do `.map(_.safe)`, I would have to, instead, have a `.map(identity)` with an implicit conversion in the result of the identity, which I find to be a more convoluted way of thinking.

I also experimented implementing the library as calls to `unsafely { _.mapValues(xxx) }`, and I'm not sure of which is a better idiom.

Reply:
> It's essentially the same you were already doing in `KStreamS` too.

I was trying to think what additional benefit `TSKType` brings on us. `KStreamS` is a plain wrapper over the Java instance and a plain old implicit in `ImplicitConversions` does all necessary conversions. Do you see any potential *unsafety* in that approach that led you to come up with the implicit classes with `safe` methods ?

Reply:
For example, take this code under consideration:

```scala
val x: KStream[K, V] = ...

implicit def KStreamToTS(instance: KStream[K, V]): TSKStream[K, V] = new TSKStream(instance)

x.mapValues((t: V) => f(t))
```

In the later expression, we will be using the KStream#mapValues method, but if we want to use the TSKStream#mapValues method, there is no easy way to fix that. However, by prepending `.safe` you can select to apply the transformation selectively if you want it.

Keeping both APIs available from the same object makes it harder to debug problems which may arise from the implicit conversions not getting always triggered, and by separating the APIs via an intermediate `.safe` step, it's easier to distinguish. Alternatively, the user could adscribe their object to the TSKStream type (e.g., ` (x : TSKStream[K, V]).mapValues(???)`) but that is more convoluted than adding `.safe`.

In comparison with KStreamS, this wrapper is a bit less *thin*, as it will try to include the appropriate Serde/Produced/etc instances, so you may want to know if you are using the TS version or not because otherwise you may need to produce the serializing objects yourself, so I find it's a bit more important to distinguish whether you're using the wrapper or not.

Reply:
For simple usage cases, such as implementing the API both methods will suffice, as you are stating the return types explicitly (and thus you are guaranteed that the conversion will happen inside the function), but if your application still uses KStreams around, it's probaly easier to keep track of which operations you are performing by adding them explicitly.

Reply:
> In the later expression, we will be using the `KStream#mapValues` method, but if we want to use the `TSKStream#mapValues` method, there is no easy way to fix that. However, by prepending `.safe` you can select to apply the transformation selectively if you want it.

This is a relatively rare use case and as u mentioned later type ascription can be used as well. So `(x: KStreamS).mapValues` can be a good workaround with the earlier implementation.

To be honest I am still not convinced about the relative pay offs of the additional types, though, as you said above, there may be some complex use cases where they offer advantages. The question is whether we should make the implementation complex enough to cater to some of the very uncommon use cases. Also can we qualify some of those use cases where this may be useful ?

@seglo, @deanwampler - what do you think ?

Reply:
I think using type ascription to hint to use the implicit conversion is something any end user with basic OO knowledge will understand.  If the call to `safe` is simply facilitating this conversion then I don't see a lot of value add.

I don't know how common the use case will be to do this conversion in user application code.  Both the existing kafka-streams-scala implementation, and this typesafe alternative, provide explicit ways to construct the stream so you wouldn't need to worry about the base types in most circumstances, but obviously having the option to cast to and back is a requirement for compatibility with the rest of the kafka streams ecosystem.  

This is just another alternate opinion to consider of many already present in this PR, but in keeping with the original intent of this library I think the tie breaker is that we should favour a thin abstraction over a heavier abstraction when they more or less satisfy the same requirement.

Reply:
I've set the change on another branch to use an `implicit def` instead of an `implicit class`, but see that in https://github.com/openshine/kafka-streams-scala/pull/1/files#diff-dfd6a81a0b5cd3214dc0e71d172fa824L36 we can no longer say `.map(_.safe)` and in order to resolve the type application we have to call the explicit type converter (or otherwise we would have to adscribe the element to the type much more verbosely, like `.map(s => s : TSKStream[K, V])`.